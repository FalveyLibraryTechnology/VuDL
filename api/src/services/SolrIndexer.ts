import Fedora from "./Fedora";
import FedoraData from "../models/FedoraData";
import HierarchyCollector from "./HierarchyCollector";

interface SolrFields {
    [key: string]: string | Array<string>;
}

class SolrIndexer {
    hierarchyCollector: HierarchyCollector;
    // TODO: make configurable
    institution = "Villanova University";
    collection = "Digital Library";

    constructor() {
        // Make Fedora connection
        const fedora = new Fedora();
        // TODO: make configurable
        const topPids = ["vudl:1", "vudl:3"];
        this.hierarchyCollector = new HierarchyCollector(fedora, topPids);
    }

    protected padNumber(num: string): string {
        // Yes, I wrote a left_pad function.
        const paddedNumber = "0000000000" + num;
        return paddedNumber.substr(paddedNumber.length - 10);
    }

    async getFields(pid: string): Promise<SolrFields> {
        // Collect hierarchy data
        const fedoraData = await this.hierarchyCollector.getHierarchy(pid);

        // Start with basic data:
        const fields: SolrFields = {
            id: pid,
            record_format: "vudl",
            institution: this.institution,
            collection: this.collection,
            modeltype_str_mv: fedoraData.models,
            datastream_str_mv: fedoraData.fedoraDatastreams,
            hierarchytype: "",
            hierarchy_all_parents_str_mv: fedoraData.getAllParents(),
        };

        // Load RELS-EXT data (some of this is used below):
        for (const field in fedoraData.relations) {
            const fieldName = "relsext." + field + "_txt_mv";
            fields[fieldName] = fedoraData.relations[field];
        }

        // Is this a hierarchy?
        if (fedoraData.models.includes("vudl-system:FolderCollection")) {
            fields.is_hierarchy_id = fedoraData.pid;
            fields.is_hierarchy_title = fedoraData.title;
        }

        // Add sequence/order data:
        const sequenceIndex = {};
        for (const sequence of fedoraData.sequences) {
            const [seqPid, seqNum] = sequence.split("#", 2);
            sequenceIndex[seqPid] = seqNum;
            const sequence_str = seqPid.replace(":", "_");
            const dynamic_sequence_field_name = "sequence_" + sequence_str + "_str";
            fields[dynamic_sequence_field_name] = this.padNumber(seqNum);
        }

        // Process parent data:
        const hierarchyParents: Array<FedoraData> = [];
        const hierarchySequences: Array<string> = [];
        for (const parent of fedoraData.parents) {
            // If the object is a Data, the parentPID is the Resource it belongs
            // to (skip the List object):
            if (fedoraData.models.includes("vudl-system:DataModel")) {
                for (const grandParent of parent.parents) {
                    hierarchyParents.push(grandParent);
                    hierarchySequences.push(this.padNumber(sequenceIndex[grandParent.pid] ?? 0));
                }
            } else {
                // ...else it is the immediate parent (Folder most likely):
                hierarchyParents.push(parent);
                hierarchySequences.push(this.padNumber(sequenceIndex[parent.pid] ?? 0));
            }
        }
        const hierarchyTops: Array<FedoraData> = fedoraData.getAllHierarchyTops();
        if (hierarchyTops.length > 0) {
            fields.hierarchy_top_id = [];
            fields.hierarchy_top_title = [];
            for (const top of hierarchyTops) {
                if (!fields.hierarchy_top_id.includes(top.pid)) {
                    fields.hierarchy_top_id.push(top.pid);
                    fields.hierarchy_top_title.push(top.title);
                }
            }
        }
        if (hierarchyParents.length > 0) {
            // This is what we are collapsing on:
            fields.hierarchy_first_parent_id_str = fedoraData.models.includes("vudl-system:DataModel")
                ? hierarchyParents[0].pid
                : pid;
            fields.hierarchy_browse = [];
            fields.hierarchy_parent_id = [];
            fields.hierarchy_parent_title = [];
            for (const parent of hierarchyParents) {
                if (!fields.hierarchy_parent_id.includes(parent.pid)) {
                    fields.hierarchy_browse.push(parent.title + "{{{_ID_}}}" + parent.pid);
                    fields.hierarchy_parent_id.push(parent.pid);
                    fields.hierarchy_parent_title.push(parent.title);
                }
            }
        } else {
            // If no parents, we still need to include the current object as
            // its own parent for field-collapsing purposes, and we can figure
            // out which top-level container is in play using RELS data.
            fields.hierarchy_first_parent_id_str = pid;
            if (typeof fields["relsext.isMemberOf_txt_mv"] !== "undefined") {
                fields.hierarchy_parent_id = fields["relsext.isMemberOf_txt_mv"];
            }
            fields.hierarchy_sequence = this.padNumber("0");
            // TODO: is hierarchy_parent_title needed in this situation?
        }
        if (hierarchySequences.length > 0) {
            fields.hierarchy_sequence_sort_str = hierarchySequences[0] ?? this.padNumber("0");
            fields.hierarchy_sequence = hierarchySequences;
        }

        // Load all the Dublin Core data:
        for (const field in fedoraData.metadata) {
            const fieldName = field.replace(":", ".") + "_txt_mv";
            fields[fieldName] = fedoraData.metadata[field];
        }

        // This map copies existing values as-is to other fields:
        const copyFields = {
            author: "dc.creator_txt_mv",
            author2: "dc.contributor_txt_mv",
            dc_source_str_mv: "dc.source_txt_mv",
            description: "dc.description_txt_mv",
            format: "dc.format_txt_mv",
            publisher: "dc.publisher_txt_mv",
            publisher_str_mv: "dc.publisher_txt_mv",
            series: "dc.relation_txt_mv",
            topic: "dc.subject_txt_mv",
            topic_facet: "dc.subject_txt_mv",
            topic_str_mv: "dc.subject_txt_mv",
        };
        for (const field in copyFields) {
            if (typeof fields[copyFields[field]] !== "undefined") {
                fields[field] = fields[copyFields[field]];
            }
        }

        // This map copies the first value from existing fields to
        // new fields:
        const firstOnlyFields = {
            author_sort: "dc.creator_txt_mv",
            dc_date_str: "dc.date_txt_mv",
            dc_relation_str: "dc.relation_txt_mv",
            dc_title_str: "dc.title_txt_mv",
            title: "dc.title_txt_mv",
            title_full: "dc.title_txt_mv",
            title_short: "dc.title_txt_mv",
            title_sort: "dc.title_txt_mv",
        };
        for (const field in firstOnlyFields) {
            if (typeof fields[firstOnlyFields[field]] !== "undefined") {
                fields[field] = fields[firstOnlyFields[field]][0];
            }
        }

        // This map copies all values AFTER the first to new fields:
        const secondaryValueFields = {
            title_alt: "dc.title_txt_mv",
        };
        for (const field in secondaryValueFields) {
            if (
                typeof fields[secondaryValueFields[field]] !== "undefined" &&
                fields[secondaryValueFields[field]].length > 1
            ) {
                fields[field] = fields[secondaryValueFields[field]].slice(1);
            }
        }

        // If this is a data model, we want to pull the date from its parent.
        const dateString = fedoraData.models.includes("vudl-system:DataModel")
            ? hierarchyParents[0].metadata["dc:date"][0] ?? ""
            : (fields["dc.date_txt_mv"] ?? [])[0] ?? "";
        const strippedDate = dateString.substr(0, 4);
        // TODO: configurable date cut-off?
        if (parseInt(strippedDate) > 1000) {
            fields.publishDate = strippedDate;
            fields.publishDateSort = strippedDate;
            // TODO: fields.normalized_sort_date = this.normalizeSortDate(dateString);
        }

        // TODO: configurable/complete language map:
        const languageMap = { en: "English" };
        if (typeof fields["dc.language_txt_mv"] !== "undefined") {
            fields.language = (fields["dc.language_txt_mv"] as Array<string>).map((lang) => {
                return languageMap[lang] ?? lang;
            });
        }

        if (typeof fields["title"] !== "undefined") {
            // TODO: configurable article list:
            const articles = ["a ", "an ", "the "];
            let sortTitle = (fields["title"] as string).toLowerCase();
            for (const article of articles) {
                if (sortTitle.substr(0, article.length) === article) {
                    sortTitle = sortTitle.substr(article.length);
                    break;
                }
            }
            fields.title_sort = sortTitle.trim();
            if (!fedoraData.models.includes("vudl-system:DataModel")) {
                fields.collection_title_sort_str = fields.title_sort;
            }
        }

        fields.has_order_str = ((fields["relsext.sortOn_txt_mv"] ?? [])[0] ?? "title") === "custom" ? "yes" : "no";

        for (const field in fedoraData.fedoraDetails) {
            const fieldName = "fgs." + field + "_txt_mv";
            fields[fieldName] = fedoraData.fedoraDetails[field];
        }

        // TODO: fields[license.mdRef]
        // TODO: fields[agent.name]
        // TODO: fields.has_thumbnail
        // TODO: fields.THUMBNAIL_contentDigest_type
        // TODO: fields.THUMBNAIL_contentDigest_digest
        // TODO: fields.THUMBNAIL_contentLocation_type
        // TODO: fields.THUMBNAIL_contentLocation_ref

        return fields;
    }
}

export default SolrIndexer;
