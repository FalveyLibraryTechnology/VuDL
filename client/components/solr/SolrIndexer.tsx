import React, { useState } from "react";
import SinglePidIndexer from "./SinglePidIndexer";
import PidRangeIndexer from "./PidRangeIndexer";
import Accordion from "@mui/material/Accordion";
import AccordionSummary from "@mui/material/AccordionSummary";
import AccordionDetails from "@mui/material/AccordionDetails";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";

const SolrIndexer = (): React.ReactElement => {
    const [results, setResults] = useState("");

    return (
        <div>
            <h1>Solr Index Tool</h1>
            <Accordion>
                <AccordionSummary expandIcon={<ExpandMoreIcon />}>Index Single PID:</AccordionSummary>
                <AccordionDetails>
                    <SinglePidIndexer setResults={setResults} />
                </AccordionDetails>
            </Accordion>
            <Accordion>
                <AccordionSummary expandIcon={<ExpandMoreIcon />}>Index Range of PIDs:</AccordionSummary>
                <AccordionDetails>
                    <PidRangeIndexer setResults={setResults} />
                </AccordionDetails>
            </Accordion>
            <h2>Results:</h2>
            <pre id="solrIndexResults">{results}</pre>
        </div>
    );
};

export default SolrIndexer;
