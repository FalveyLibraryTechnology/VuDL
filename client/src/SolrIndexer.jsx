import React from "react";
import AjaxHelper from "./AjaxHelper";

export default function SolrIndexer() {
    function doApiCall(method) {
        const ajax = AjaxHelper.getInstance();
        const pid = document.getElementById("solrIndexPid").value;
        const url = ajax.url + "/messenger/solrindex/" + encodeURIComponent(pid);
        ajax.ajax({
            method: method,
            url: url,
            dataType: "text",
            error: function (result, status) {
                document.getElementById("solrIndexResults").innerHTML = status;
            },
            success: function (result) {
                document.getElementById("solrIndexResults").innerHTML = result;
            },
        });
    }

    function preview() {
        doApiCall("GET");
    }

    function index() {
        doApiCall("POST");
    }

    return (
        <div>
            <h1>Solr Index Tool</h1>
            <label>
                PID: <input type="text" id="solrIndexPid" />
            </label>
            <button onClick={preview}>Preview</button>
            <button onClick={index}>Index</button>
            <h2>Results:</h2>
            <pre id="solrIndexResults"></pre>
        </div>
    );
}
