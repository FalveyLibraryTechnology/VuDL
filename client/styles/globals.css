:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
