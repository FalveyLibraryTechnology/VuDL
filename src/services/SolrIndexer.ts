import Fedora from './Fedora';

class SolrIndexer {
    fedora: Fedora;

    constructor() {
        // TODO: Config
        // Make Fedora connection
        this.fedora = new Fedora();
    }

    async getFields(pid: string) {
        // Use Fedora to get data
        // TODO: type
        let dc = await this.fedora.getDC(pid);

        // Massage data
        let fields: any = {};

        // TODO: Pull from config
        let fieldMap = {
            "dc:identifier": "id",
            "dc:title": "title",
            "dc:subject": "subject"
        };


        for (let field of dc.children) {
            switch (field.name) {
                case "dc:date":
                    fields.date = field.value;
                    break;
                case "dc:creator":
                    fields.creator = field.value;
                    break;
                case "dc:language":
                    fields.language = field.value;
                    break;
                case "dc:relation":
                    fields.relation = field.value;
                    break;
                case "dc:source":
                    fields.source = field.value;
                    break;
                case "dc:collection":
                    fields.collection = field.value;
                    break;
                case "dc:format":
                    fields.format = field.value;
                    break;
                default:
                    if (typeof fieldMap[field.name] != "undefined") {
                        fields[fieldMap[field.name]] = field.value; 
                    }
                    break;
            }
    }

        return fields;
    }
}

export default SolrIndexer;