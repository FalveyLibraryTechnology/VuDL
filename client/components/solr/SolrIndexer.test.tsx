import React from "react";
import { describe, expect, it, jest } from "@jest/globals";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import SolrIndexer from "./SolrIndexer";

jest.mock("./SinglePidIndexer", () => () => "SinglePidIndexer");
jest.mock("./PidRangeIndexer", () => () => "PidRangeIndexer");

const mockUseFetchContext = jest.fn();
jest.mock("../../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));

describe("SolrIndexer", () => {
    it("renders", () => {
        const wrapper = mount(<SolrIndexer />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
