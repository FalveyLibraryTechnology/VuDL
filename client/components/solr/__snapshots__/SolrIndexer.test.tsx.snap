// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SolrIndexer renders 1`] = `
<div>
  BasicBreadcrumbs
  <h1>
    Solr Index Tool
  </h1>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
  >
    <div
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        Index Single PID:
      </div>
      <div
        className="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style={
        {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            className="MuiAccordion-region"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              SinglePidIndexer
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
  >
    <div
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
      >
        Index Range of PIDs:
      </div>
      <div
        className="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
      style={
        {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            className="MuiAccordion-region"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              PidRangeIndexer
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <h2>
    Results:
  </h2>
  <pre
    id="solrIndexResults"
  />
</div>
`;
